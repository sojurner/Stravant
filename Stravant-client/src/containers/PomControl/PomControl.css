.pom-instruction {
  position: absolute; }

img {
  cursor: pointer; }

i {
  cursor: pointer; }

.modal-timer {
  z-index: 1;
  background-color: #fed2bf;
  padding: 1rem;
  position: absolute;
  text-align: center;
  animation: fadein 1s; }

@keyframes fadein {
  from {
    opacity: 0; }
  to {
    opacity: 1; } }

i {
  font-size: 2rem; }
