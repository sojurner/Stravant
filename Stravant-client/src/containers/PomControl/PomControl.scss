.pom-instruction {
}

img {
  cursor: pointer;
}
i {
  cursor: pointer;
}

.modal-timer {
  z-index: 1;
  background-color: #fed2bf;
  padding: 1rem;
  position: absolute;
  text-align: center;
  animation: fadein 1s;
}

.fa-hand-paper {
  font-size: 4rem;
}

.fa-hand-paper:hover {
  color: rgb(131, 52, 52);
}

@keyframes fadein {
  from {
    opacity: 0;
  }
  to {
    opacity: 1;
  }
}

i {
  font-size: 2rem;
}
