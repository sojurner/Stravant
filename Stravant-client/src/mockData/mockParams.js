export const tokenParams = [
  'https://www.strava.com/oauth/token',
  {
    body:
      '{"client_id":28236,"client_secret":"829f12d9bdb8ce424edf87deff4551c117d3ed36","code":"r5423rffda4c234ctwtwert2ct"}',
    headers: { 'Content-Type': 'application/json' },
    method: 'POST'
  }
];

export const aggregateParams = [
  'https://www.strava.com/api/v3/athletes/28236/stats',
  {
    headers: {
      Authorization: 'Bearer a7eeea216a46a760a74c60acbffc3b55c66537c5',
      'Content-Type': 'application/json'
    },
    method: 'GET'
  }
];

export const weeklyParams = [
  'https://www.strava.com/api/v3/athlete/activities?before=1536123319&after=1536036919',
  {
    headers: {
      Authorization: 'Bearer a7eeea216a46a760a74c60acbffc3b55c66537c5',
      'Content-Type': 'application/json'
    },
    method: 'GET'
  }
];
