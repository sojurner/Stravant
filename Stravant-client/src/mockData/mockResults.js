export const aggregateResult = { Biked: 0, Ran: 6.472964574269732, Swam: 0 };

export const tokenResult = {
  access_token: 'a7eeea216a46a760a74c60acbffc3b55c66537c5',
  athlete: {
    badge_type_id: 0,
    city: null,
    country: null,
    created_at: '2018-09-04T22:53:48Z',
    email: 'pykim0591@gmail.com',
    firstname: 'Paul',
    follower: null,
    friend: null,
    id: 34515396,
    lastname: 'K',
    premium: false,
    profile:
      'https://lh4.googleusercontent.com/-q79bv4Nc_dY/AAAAAAAAAAI/AAAAAAAAAAA/APUIFaMcxPJyTx-ZbjWs8aCsKgkL96lV4w/mo/photo.jpg',
    profile_medium:
      'https://lh4.googleusercontent.com/-q79bv4Nc_dY/AAAAAAAAAAI/AAAAAAAAAAA/APUIFaMcxPJyTx-ZbjWs8aCsKgkL96lV4w/mo/photo.jpg',
    resource_state: 2,
    sex: 'M',
    state: null,
    summit: false,
    updated_at: '2018-09-07T12:20:05Z',
    username: null
  },
  token_type: 'Bearer'
};

export const weeklyResult = {
  Friday: 3784.3,
  Monday: 0,
  Saturday: 0,
  Sunday: 0,
  Thursday: 0,
  Tuesday: 0,
  Wednesday: 0
};
