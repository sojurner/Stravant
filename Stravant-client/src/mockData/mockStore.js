export const currentUser = {
  info: {
    accessToken: 'a7eeea216a46a760a74c60acbffc3b55c66537c5',
    userId: 34515396,
    firstName: 'Paul',
    gender: 'M',
    profilePic:
      'https://lh4.googleusercontent.com/-q79bv4Nc_dY/AAAAAAAAAAI/AAAAAAAAAAA/APUIFaMcxPJyTx-ZbjWs8aCsKgkL96lV4w/mo/photo.jpg'
  },
  totalStats: {},
  weeklyStats: {}
};

export const pomStatus = {
  pomStatus: true,
  lastPom: '',
  pomHistory: {
    'Tue, Sep 11, 2018 8:39 PM': { second: 19, minute: 0, hour: 0 },
    'Tue, Sep 11, 2018 8:40 PM': { second: 3, minute: 0, hour: 0 },
    'Tue, Sep 11, 2018 10:22 PM': { second: 3, minute: 0, hour: 0 }
  }
};

export const clubInfo = {
  clubInfo: { clubName: 'Stravant Club', clubId: 473825 },
  clubActivity: {
    mostRecent: {
      name: 'Pamela',
      distance: 4.270478558110628,
      activity: 'Morning Run'
    },
    Pamela: {
      totalDistance: 62.88999999999999,
      totalTime: 635,
      membersRecent: 'Afternoon Run'
    },
    Paul: { totalDistance: 6.46, totalTime: 78, membersRecent: 'Morning Run' },
    Jesse: {
      totalDistance: 277.78000000000003,
      totalTime: 5292,
      membersRecent: 'Morning Run'
    }
  }
};

export const mockStore = {
  ...currentUser,
  ...pomStatus,
  ...clubInfo
};
