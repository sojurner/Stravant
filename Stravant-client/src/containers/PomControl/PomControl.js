import React, { Component } from 'react';
import { connect } from 'react-redux';
import openSocket from 'socket.io-client';
import moment from 'moment';
// import { createActivity } from '../../helpers/apiCalls/apiCalls';
import { PomContainer } from '../../components/PomContainer/PomContainer';
import * as pomActions from '../../actions/pomAction';
import { getLastPomTime } from '../../helpers/helpers/helpers';

import './PomControl.css';

var socket = openSocket('http://localhost:5000');
export class PomControl extends Component {
  constructor() {
    super();
    this.state = {
      mSecond: 0,
      second: 0,
      minute: 0,
      hour: 0,
      start: false,
      stop: true,
      save: false,
      hide: false,
      description: false,
      pomSummary: '',
      socketMessage: ''
    };
  }

  componentDidMount() {
    const { setPomHistory } = this.props;
    if (localStorage.getItem('pomHistory')) {
      const storageItem = JSON.parse(localStorage.getItem('pomHistory'));
      setPomHistory(storageItem);
    }
  }

  togglePom = (event, string) => {
    const { start, hide } = this.state;
    if (start) {
      const toggleHide = !hide;
      this.setState({ hide: toggleHide });
    } else {
      this.sendSocketMsg(event);
      this.setState({ [string]: true, stop: false });
    }
    const { togglePomState, pomInfo } = this.props;
    togglePomState(!pomInfo.pomStatus);
  };

  sendSocketMsg = () => {
    const { firstName } = this.props.currentUser.info;
    socket.emit('pomStart', `${firstName} just started on Pom!`);
  };

  startMSecond = () => {
    let time = this.state.mSecond + 1;
    time < 10
      ? this.setState({ mSecond: time })
      : this.setState({ mSecond: 0 });
  };

  startSecond = () => {
    let time = this.state.second + 1;
    time < 60 ? this.setState({ second: time }) : this.setState({ second: 0 });
  };

  startMinute = () => {
    let time = this.state.minute + 1;
    time < 60 ? this.setState({ minute: time }) : this.setState({ minute: 0 });
  };

  startHour = () => {
    let time = this.state.hour + 1;
    this.setState({ hour: time });
  };

  stopTime = () => {
    const { firstName } = this.props.currentUser.info;
    const { pomHistory } = this.props.pomInfo;
    const timeKeys = Object.keys(pomHistory);
    const lastPom = timeKeys[timeKeys.length - 1];
    const currentTime = moment().format('llll');
    if (lastPom) {
      const difference = getLastPomTime(currentTime, lastPom);
      socket.emit(
        'pomEnd',
        `${firstName} finished the Pom!  Your last pom was ${difference}`
      );
    }
    this.setState({ start: false, stop: true });
  };

  addDescription = string => {
    this.setState({ description: string });
  };

  removeDescription = () => {
    this.setState({ description: '' });
  };

  resetTimer = () => {
    const { setPomHistory } = this.props;
    const { minute, second, hour } = this.state;
    const timeState = {
      second: second,
      minute: minute,
      hour: hour
    };
    if (localStorage.getItem('pomHistory')) {
      const storageItem = JSON.parse(localStorage.getItem('pomHistory'));
      const newItem = {
        ...storageItem,
        [moment().format('llll')]: timeState
      };
      localStorage.setItem('pomHistory', JSON.stringify(newItem));
    } else {
      const itemToStorage = {
        [moment().format('llll')]: timeState
      };
      localStorage.setItem('pomHistory', JSON.stringify(itemToStorage));
    }
    const storageItem = JSON.parse(localStorage.getItem('pomHistory'));
    setPomHistory(storageItem);

    const timeSummary = `${second}s  ${minute}m  ${hour}h`;
    this.setState({
      mSecond: 0,
      second: 0,
      minute: 0,
      hour: 0,
      start: false,
      save: true,
      pomSummary: timeSummary
    });
  };

  removePom = (history, time) => {
    delete history[time];
    this.props.setPomHistory(history);
    localStorage.setItem('pomHistory', JSON.stringify(history));
  };

  render() {
    socket.on('pomClick', msg => {
      this.setState({ socketMessage: msg });
    });
    const { pomHistory } = this.props.pomInfo;
    const styles = {
      pomStyle: {
        position: 'absolute',
        left: window.screenX + 10,
        top: window.screenY - 20
      },

      summaryStyle: {
        position: 'absolute',
        left: window.screenX + 25,
        top: window.screenY + 90
      },

      suggestStyle: {
        position: 'absolute',
        left: window.screenX + 25,
        top: window.screenY + 60
      },

      modalStyle: {
        position: 'absolute',
        left: window.screenX + 100,
        top: window.screenY - 20
      }
    };
    const {
      mSecond,
      second,
      minute,
      hour,
      start,
      stop,
      description,
      hide,
      save,
      pomSummary,
      socketMessage
    } = this.state;
    return (
      <div>
        {(second > 0 || mSecond > 0 || hour > 0 || minute > 0) &&
          !hide &&
          start && (
            <section className="modal" style={styles.modalStyle}>
              <div className="time">
                <span className="time hour">{hour} h</span>
                <span className="time minute">
                  ..
                  {minute} m
                </span>
                <span className="time second">
                  ..
                  {second} s
                </span>
                <span className="time mSecond">
                  ..
                  {mSecond} ms
                </span>
              </div>
              <div className="save-stop">
                <i
                  class="far fa-hand-paper"
                  onClick={() => {
                    this.stopTime();
                    if (!stop) {
                      clearInterval(this.mSeconds);
                      clearInterval(this.seconds);
                      clearInterval(this.minutes);
                      clearInterval(this.hours);
                      this.resetTimer();
                    }
                  }}
                  onMouseEnter={() => this.addDescription('stop')}
                  onMouseLeave={this.removeDescription}
                />
              </div>
            </section>
          )}
        {save && (
          <h4 style={styles.summaryStyle}>
            <div>Most recent:</div>
            ---
            {pomSummary}
            ---
          </h4>
        )}
        {socketMessage && <h4>{socketMessage}</h4>}
        <img
          src={require('../../images/pomodoro-icon.png')}
          style={styles.pomStyle}
          height="100px"
          width="100px"
          className="pomodoro"
          onMouseEnter={() => this.addDescription('start')}
          onMouseLeave={this.removeDescription}
          onClick={event => {
            this.togglePom(event, 'start');
            if (!start && stop) {
              this.mSeconds = setInterval(this.startMSecond, 101);
              this.seconds = setInterval(this.startSecond, 1001);
              this.minutes = setInterval(this.startMinute, 60001);
              this.hours = setInterval(this.startHour, 3600001);
            }
          }}
        />
        {!pomHistory && <p>You have no record Poms</p>}
        {description === 'start' &&
          !hide &&
          !start && (
            <p style={styles.suggestStyle} className="pom-instruction">
              Start Pom?
            </p>
          )}
        {description === 'start' &&
          !hide &&
          start && (
            <p style={styles.suggestStyle} className="pom-instruction">
              Hide Pom?
            </p>
          )}
        {description === 'stop' && (
          <p style={styles.suggestStyle} className="stop-instruction">
            End Pom?
          </p>
        )}

        {hide && (
          <p style={styles.suggestStyle} className="stop-instruction">
            Show Pom?
          </p>
        )}
        <PomContainer removePom={this.removePom} pomHistory={pomHistory} />
      </div>
    );
  }
}

const mapStateToProps = state => ({
  currentUser: state.currentUser,
  pomInfo: state.pomInfo
});

const mapDispatchToProps = dispatch => ({
  setPomHistory: history => dispatch(pomActions.setPomHistory(history)),
  togglePomState: bool => dispatch(pomActions.togglePomState(bool))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(PomControl);
