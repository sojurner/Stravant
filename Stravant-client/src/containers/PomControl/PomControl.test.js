import React from 'react';
import { PomControl, mapStateToProps, mapDispatchToProps } from './PomControl';
import * as mock from './mockData';
import * as store from '../../mockData/mockStore';

describe('PomControl', () => {
  let wrapper;
  let state;
  let defaultState;
  let mockSetPom;
  let mockToggle;

  beforeEach(() => {
    localStorage.clear();
    mockSetPom = jest.fn();
    mockToggle = jest.fn();
    wrapper = shallow(
      <PomControl
        currentUser={mock.currentUser}
        pomInfo={mock.pomInfo}
        setPomHistory={mockSetPom}
        togglePomState={mockToggle}
      />
    );
  });

  defaultState = {
    description: false,
    hide: false,
    hour: 0,
    mSecond: 0,
    minute: 0,
    pomSummary: '7s  2m  1h',
    save: true,
    second: 0,
    socketMessage: '',
    start: false,
    stop: true
  };

  state = {
    mSecond: 4,
    second: 7,
    minute: 2,
    hour: 1,
    save: false
  };

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should setState of save to false and togglePomState to have been called', () => {
    expect(wrapper.state('start')).toEqual(false);

    wrapper.instance().togglePom({}, 'start');

    expect(wrapper.state('start')).toEqual(true);
    expect(wrapper.state('stop')).toEqual(false);

    wrapper.instance().togglePom({}, 'hide');

    expect(wrapper.state('hide')).toEqual(true);
    expect(mockToggle).toHaveBeenCalled();
  });

  it('should increment state of mSecond by 1 and reset to 0 when it reaches 10', () => {
    wrapper.setState({ mSecond: 8 });

    wrapper.instance().startMSecond();

    expect(wrapper.state('mSecond')).toEqual(9);

    wrapper.instance().startMSecond();

    expect(wrapper.state('mSecond')).toEqual(0);
  });

  it('should increment state of second by 1 and reset to 0 when it reaches 60', () => {
    wrapper.setState({ second: 58 });

    wrapper.instance().startSecond();

    expect(wrapper.state('second')).toEqual(59);

    wrapper.instance().startSecond();

    expect(wrapper.state('second')).toEqual(0);
  });

  it('should increment state of Minute by 1 and reset to 1 when it reaches 60', () => {
    wrapper.setState({ minute: 58 });

    wrapper.instance().startMinute();

    expect(wrapper.state('minute')).toEqual(59);

    wrapper.instance().startMinute();

    expect(wrapper.state('minute')).toEqual(0);
  });

  it('should increment state of Hour by 1', () => {
    wrapper.setState({ hour: 0 });

    wrapper.instance().startHour();

    expect(wrapper.state('hour')).toEqual(1);

    wrapper.instance().startHour();

    expect(wrapper.state('hour')).toEqual(2);
  });

  it('should stop time and set State of start and stop', () => {
    wrapper.setState({ start: true, stop: false });
    wrapper.instance().stopTime();

    expect(wrapper.state('start')).toEqual(false);
    expect(wrapper.state('stop')).toEqual(true);
  });

  it('should reset the state of times to true and save to 0', () => {
    wrapper.setState(state);

    setTimeout(() => {
      wrapper.instance().resetTimer();
    });

    expect(wrapper.state('pomSummary')).toEqual('');
  });

  it('should set state of description whn addDescription is called', () => {
    wrapper.instance().addDescription('paul');

    expect(wrapper.state('description')).toEqual('paul');
  });

  it('should set state of description to an empty string when removeDescription is called', () => {
    wrapper.setState({ description: 'paul' });

    wrapper.instance().removeDescription();
    expect(wrapper.state('description')).toEqual('');
  });

  it('should grab from local storage', () => {
    const local = {
      'Mon, Sep 10, 2018 9:12 PM': {
        mSecond: '0',
        second: '0',
        minute: '0',
        hour: '0',
        save: 'true',
        showHistory: 'true',
        description: 'false'
      }
    };
    const stringifiedLocal = JSON.stringify(local);
    let expectedLocal = localStorage.store;

    wrapper.setState(state);
    localStorage.setItem('pomHistory', stringifiedLocal);
    wrapper.instance().resetTimer();
    expect(mockSetPom).toHaveBeenCalled();
  });

  it('should set new item to storage if there is no pom history', () => {
    const expected = localStorage.store;

    wrapper.setState({ minute: 4, second: 5, hour: 10 });
    wrapper.instance().resetTimer();

    expect(localStorage.store).toEqual(expected);
    expect(wrapper.state()).toEqual({
      description: false,
      hide: false,
      hour: 0,
      mSecond: 0,
      minute: 0,
      pomSummary: '5s  4m  10h',
      save: true,
      second: 0,
      socketMessage: '',
      start: false,
      stop: true
    });
  });

  it('should remove pom from store', () => {
    const local = {
      'Mon, Sep 10, 2018 9:12 PM': {
        second: '1',
        minute: '2',
        hour: '3'
      },
      'Mon, Sep 15, 2017 9:12 PM': {
        second: '4',
        minute: '3',
        hour: '1'
      }
    };
    const expected = {
      'Mon, Sep 15, 2017 9:12 PM': {
        second: '4',
        minute: '3',
        hour: '1'
      }
    };
    const stringifiedLocal = JSON.stringify(local);
    const stringifiedExpected = JSON.stringify(expected);

    localStorage.setItem('pomHistory', stringifiedLocal);

    wrapper.instance().removePom(local, 'Mon, Sep 10, 2018 9:12 PM');
    expect(localStorage.store).toEqual({ pomHistory: stringifiedExpected });
    expect(mockSetPom).toHaveBeenCalled();
    expect(wrapper.state('pomSummary')).toEqual('');
  });

  it('should start the timer on click', () => {
    const spy = jest.spyOn(wrapper.instance(), 'togglePom');
    wrapper.instance().forceUpdate();
    const movieCard = wrapper.find('i');
    movieCard.simulate('click');
    expect(spy).toHaveBeenCalled();
  });

  it('should add Description on hover', () => {
    const spy = jest.spyOn(wrapper.instance(), 'addDescription');
    wrapper.instance().forceUpdate();
    const movieCard = wrapper.find('i');
    movieCard.simulate('mouseEnter');
    expect(spy).toHaveBeenCalled();
  });

  it('should remove description on hoveroff', () => {
    const spy = jest.spyOn(wrapper.instance(), 'removeDescription');
    wrapper.instance().forceUpdate();
    const movieCard = wrapper.find('i');
    movieCard.simulate('mouseLeave');
    expect(spy).toHaveBeenCalled();
  });

  it('should match snapshot when state of time changes', () => {
    wrapper.setState({ second: 1, hide: false, start: true, stop: false });
    wrapper.update();

    expect(wrapper).toMatchSnapshot();

    const spyStop = jest.spyOn(wrapper.instance(), 'stopTime');
    const spyReset = jest.spyOn(wrapper.instance(), 'resetTimer');
    wrapper.instance().forceUpdate();
    wrapper
      .find('i')
      .at(1)
      .simulate('click');
    expect(spyStop).toHaveBeenCalled();
    expect(spyReset).toHaveBeenCalled();
  });

  it('should add a stop description on mouse enter', () => {
    wrapper.setState({ second: 1, hide: false, start: true });
    wrapper.update();

    const spyEnter = jest.spyOn(wrapper.instance(), 'addDescription');
    wrapper.instance().forceUpdate();
    wrapper
      .find('i')
      .at(1)
      .simulate('mouseEnter');
    expect(spyEnter).toHaveBeenCalled();

    const spyLeave = jest.spyOn(wrapper.instance(), 'removeDescription');
    wrapper.instance().forceUpdate();
    wrapper
      .find('i')
      .at(1)
      .simulate('mouseLeave');
    expect(spyLeave).toHaveBeenCalled();
  });

  it('should match snapshot when description is start hide is false and start is false', () => {
    wrapper.setState({ description: 'start', hide: false, start: false });
    wrapper.update();

    expect(wrapper).toMatchSnapshot();
  });

  it('should map to proper store', () => {
    let mockStore = store.currentUser;
    let mapped = mapStateToProps(mockStore);

    expect(mapped.currentUser).toEqual(mockStore.currentUser);
  });

  it('should map to proper store', () => {
    let mockStore = store.pomStatus;
    let mapped = mapStateToProps(mockStore);

    expect(mapped.pomInfo).toEqual(mockStore.pomInfo);
  });

  it('should call dispatch method when a function from map dispatch to props is called', () => {
    const mockDispatch = jest.fn();
    const mapped = mapDispatchToProps(mockDispatch);

    mapped.setPomHistory();

    expect(mockDispatch).toHaveBeenCalled();
  });

  it('should call dispatch method when a function from map dispatch to props is called', () => {
    const mockDispatch = jest.fn();
    const mapped = mapDispatchToProps(mockDispatch);

    mapped.togglePomState();

    expect(mockDispatch).toHaveBeenCalled();
  });
});
