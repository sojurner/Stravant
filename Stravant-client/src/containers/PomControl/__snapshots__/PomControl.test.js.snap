// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PomControl should match snapShot 1`] = `
<div>
  <img
    className="pomodoro"
    height="100px"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    src="pomodoro-icon.png"
    style={
      Object {
        "left": 10,
        "position": "absolute",
        "top": -20,
      }
    }
    width="100px"
  />
  <button
    onClick={[Function]}
  >
    Show Poms
  </button>
</div>
`;

exports[`PomControl should match snapshot 1`] = `
<div>
  <img
    className="pomodoro"
    height="100px"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    src="pomodoro-icon.png"
    style={
      Object {
        "left": 10,
        "position": "absolute",
        "top": -20,
      }
    }
    width="100px"
  />
  <button
    onClick={[Function]}
  >
    Show Poms
  </button>
</div>
`;
